### PR TITLE
Add OSS support

### LINKED ISSUE #41: AlibabaCloud OSS support
This feature request is from  tensorflow/tensorflow#24556 (/cc @oss-developer):

AlibabaCloud Object Storage Service(OSS) is one of the most widely used cloud stroage services in the world. Could I ask for OSS support with tensorflow?

### PR DESCRIPTION
**Pull request by zou000**
Fix #41

### CODE REVIEW

**Comment by yongtang** [comment]
@zou000 Can you move the python test file to `tests/` directory? In this way it could be automatically picked up by pytest.

File: tensorflow_io/oss/kernels/ossfs/oss_file_system.cc [hunk 1]
```diff
@@ -0,0 +1,1158 @@
...
+Status OSSFileSystem::GetMatchingPaths(const std::string& pattern,
+                                       std::vector<std::string>* results) {
+  results->clear();
+  // Find the fixed prefix by looking for the first wildcard.
+  const string& fixed_prefix =
+      pattern.substr(0, pattern.find_first_of("*?[\\"));
+  const string& dir = string(io::Dirname(fixed_prefix));
+  const string& base_prefix = string(io::Basename(fixed_prefix));
+  if (dir.empty()) {
+    return errors::InvalidArgument("A oss pattern doesn't have a bucket name: ",
```
Comment (terrytangyuan) [inline tensorflow_io/oss/kernels/ossfs/oss_file_system.cc]: nit: "An OSS "

File: tensorflow_io/oss/kernels/ossfs/oss_file_system.cc [hunk 2]
```diff
@@ -0,0 +1,1158 @@
...
+  std::vector<std::string> listing;
+  TF_RETURN_IF_ERROR(_ListObjects(pool, oss_options, bucket, object, &listing,
+                                  true, false, 1000));
+
+  // Match all obtained paths to the input pattern.
+  for (auto path : listing) {
+    if (str_util::StartsWith(StringPiece(path), base_prefix)) {
+      const string file_path = io::JoinPath(dir, path);
+      if (Env::Default()->MatchPath(file_path, pattern)) {
+        results->push_back(dir + "/" + path);
```
Comment (terrytangyuan) [inline tensorflow_io/oss/kernels/ossfs/oss_file_system.cc]: Perhaps move "/" to a constant? Also there's inconsistent use of single and double quotes.

File: tensorflow_io/oss/kernels/ossfs/oss_file_system.cc [hunk 3]
```diff
@@ -0,0 +1,1158 @@
...
+  TF_RETURN_IF_ERROR(oss_initialize());
+  std::string object, bucket;
+  TF_RETURN_IF_ERROR(_ParseOSSURIPath(dirname, &bucket, &object));
+  OSSConnection oss(host_, access_id_, access_key_);
+  oss_request_options_t* ossOptions = oss.getRequestOptions();
+  aos_pool_t* pool = oss.getPool();
+  StringPiece dirs(object);
+
+  std::vector<std::string> splitPaths =
+      str_util::Split(dirs, '/', str_util::SkipEmpty());
```
Comment (terrytangyuan) [inline tensorflow_io/oss/kernels/ossfs/oss_file_system.cc]: Consider moving this to a separate function given it appears in multiple places

File: tensorflow_io/oss/kernels/ossfs/oss_file_system.cc [hunk 4]
```diff
@@ -0,0 +1,1158 @@
...
+          ++*undeleted_dirs;
+        } else {
+          ++*undeleted_files;
+        }
+      }
+    }
+  }
+
+  if (*undeleted_dirs == 0 && *undeleted_files == 0) {
+    // delete its self
```
Comment (terrytangyuan) [inline tensorflow_io/oss/kernels/ossfs/oss_file_system.cc]: What does this comment mean?

File: tensorflow_io/oss/kernels/ossfs/oss_file_system.cc [hunk 5]
```diff
@@ -0,0 +1,1158 @@
...
+  mutex_lock lock(mu_);
+  host_.assign(config.gets(kOSSCredentialsSection, kOSSCredentialsHostKey));
+  access_id_.assign(
+      config.gets(kOSSCredentialsSection, kOSSCredentialsAccessIdKey));
+  access_key_.assign(
+      config.gets(kOSSCredentialsSection, kOSSCredentialsAccesskeyKey));
+
+  if (host_.empty() || access_id_.empty() || access_key_.empty()) {
+    VLOG(1) << filename << " does not contains full oss credentials";
+    return errors::Internal(filename, "does not contains full oss credentials");
```
Comment (terrytangyuan) [inline tensorflow_io/oss/kernels/ossfs/oss_file_system.cc]: nit: Capitalize "oss"

File: tensorflow_io/oss/README.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+Then set environment variable `OSS_CREDENTIALS` to the path of the file.
+
+In Python code, import the extension `ossfs_op` module to use the extension with `gfile`. The files and directory URI should have `oss://` prefix, followed by a bucket name, then the directory hierarchy.
+
+```
+import tensorflow_io.oss.python.ops.ossfs_ops
+from tensorflow.python.platform import gfile
+
+gfile.MkDir('oss://your_bucket_name/test_dir')
```
Comment (terrytangyuan) [inline tensorflow_io/oss/README.md]: Perhaps also briefly mention how to use this with some of the existing dataset ops even though it seems pretty trivial.

**Comment by yongtang** [comment]
If mocking is not possible, maybe some simple sanity test to make sure at least it build and runs, e.g., access some publicly accessible "hello world" file on OSS?

**Comment by zou000** [comment]
> @zou000 Can you move the python test file to `tests/` directory? In this way it could be automatically picked up by pytest.


> @zou000 Can you move the python test file to `tests/` directory? In this way it could be automatically picked up by pytest.

Done.

File: tensorflow_io/oss/kernels/ossfs/oss_file_system.cc (hunk 1 above)
Comment (zou000) [reply tensorflow_io/oss/kernels/ossfs/oss_file_system.cc]: Done

File: tensorflow_io/oss/kernels/ossfs/oss_file_system.cc (hunk 2 above)
Comment (zou000) [reply tensorflow_io/oss/kernels/ossfs/oss_file_system.cc]: What inconsistency are you talking about? In C/C++ single and doule qutoes denote different types.

File: tensorflow_io/oss/kernels/ossfs/oss_file_system.cc (hunk 3 above)
Comment (zou000) [reply tensorflow_io/oss/kernels/ossfs/oss_file_system.cc]: There are only two usages,  and extracting a function involving passing multiple parameters and i think it actually hurts the readability?

File: tensorflow_io/oss/kernels/ossfs/oss_file_system.cc (hunk 4 above)
Comment (zou000) [reply tensorflow_io/oss/kernels/ossfs/oss_file_system.cc]: fixed

File: tensorflow_io/oss/kernels/ossfs/oss_file_system.cc (hunk 5 above)
Comment (zou000) [reply tensorflow_io/oss/kernels/ossfs/oss_file_system.cc]: fixed

File: tensorflow_io/oss/README.md (hunk 6 above)
Comment (zou000) [reply tensorflow_io/oss/README.md]: Done

**Comment by zou000** [comment]
> Looks great in general. Just left a couple small comments. Perhaps this could be some follow-up work but does OSS provide any mocking utilities similar to [moto](https://github.com/spulec/moto) for S3 so we could write some simple unit tests?

> If mocking is not possible, maybe some simple sanity test to make sure at least it build and runs, e.g., access some publicly accessible "hello world" file on OSS?

I asked around, unfortunately, it seems there is neither an OSS mocking framework nor public accessible OSS files for tests.  For example the popular OSS java sdk also rely on user's own credentials for tests: https://github.com/aliyun/aliyun-oss-java-sdk/blob/master/src/test/java/com/aliyun/oss/testing/ConcurrencyTest.java#L39 I made a change to skip OSS test cases if the env variables are absent. This is not ideal but at least it can test 
the OSS extension can be imported.

File: tensorflow_io/oss/kernels/ossfs/oss_file_system.cc (hunk 2 above)
Comment (terrytangyuan) [reply tensorflow_io/oss/kernels/ossfs/oss_file_system.cc]: Think I misread it.

**Comment by terrytangyuan** [comment]
Thanks! I think it’s fine for now. Could you fix the Travis build?

**Comment by zou000** [comment]
> Thanks! I think it’s fine for now. Could you fix the Travis build?

Travis CI needs to be configured to use some specific bazel flags, as described in detail in #200.  I am not familiar with your travis setup and I am wondering if you or @yongtang can make a quick change?   Thanks.

**Comment by yongtang** [comment]
@zou000 the docker image of custom-op actually uses `standalone` mode in bazel which might create unintended consequences (something to be fixed). It relies on environment installation which could vary a lot.

In tensorflow-io, we use `sandbox` mode to make sure all dependency binaries are specified. 

I have created a patch that should fixed the build failures. Can you apply the patch:
```patch
From dbf53e2d74e3eb90de99895c9aa83a6e9b05f9d2 Mon Sep 17 00:00:00 2001
From: Yong Tang <yong.tang.github@outlook.com>
Date: Thu, 2 May 2019 14:58:54 +0000
Subject: [PATCH] Add libexpat and fixed several dependency issues

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>
---
 WORKSPACE                     | 10 ++++++++++
 third_party/libaprutil1.BUILD |  7 +++++--
 third_party/libexpat.BUILD    | 30 ++++++++++++++++++++++++++++++
 3 files changed, 45 insertions(+), 2 deletions(-)
 create mode 100644 third_party/libexpat.BUILD

diff --git a/WORKSPACE b/WORKSPACE
index f105f64..0376dea 100644
--- a/WORKSPACE
+++ b/WORKSPACE
@@ -290,6 +290,16 @@ http_archive(
     ],
 )

+http_archive(
+    name = "libexpat",
+    build_file = "//third_party:libexpat.BUILD",
+    sha256 = "574499cba22a599393e28d99ecfa1e7fc85be7d6651d543045244d5b561cb7ff",
+    strip_prefix = "libexpat-R_2_2_6/expat",
+    urls = [
+        "http://github.com/libexpat/libexpat/archive/R_2_2_6.tar.gz",
+    ],
+)
+
 http_archive(
     name = "libapr1",
     build_file = "//third_party:libapr1.BUILD",
diff --git a/third_party/libaprutil1.BUILD b/third_party/libaprutil1.BUILD
index 6545137..ad9b6ac 100644
--- a/third_party/libaprutil1.BUILD
+++ b/third_party/libaprutil1.BUILD
@@ -4,7 +4,10 @@ licenses(["notice"])  # Apache license (for libapr1)

 cc_library(
     name = "libaprutil1",
-    srcs = [
+    srcs = glob([
+        "include/private/apu_*.h",
+        "include/private/apr_*.h",
+    ]) + [
         "buckets/apr_brigade.c",
         "buckets/apr_buckets.c",
         "buckets/apr_buckets_alloc.c",
@@ -112,10 +115,10 @@ cc_library(
         "-lrt",
         "-lcrypt",
         "-lpthread",
-        "-lexpat",
     ],
     strip_include_prefix = "include",
     deps = [
         "@libapr1",
+        "@libexpat",
     ],
 )
diff --git a/third_party/libexpat.BUILD b/third_party/libexpat.BUILD
new file mode 100644
index 0000000..f0e617c
--- /dev/null
+++ b/third_party/libexpat.BUILD
@@ -0,0 +1,30 @@
+# Description:
+#   Expat library
+
+licenses(["notice"])
+
+exports_files(["COPYING"])
+
+cc_library(
+    name = "libexpat",
+    srcs = [
+        "lib/xmlparse.c",
+        "lib/xmlrole.c",
+        "lib/xmltok.c",
+    ],
+    hdrs = glob([
+        "lib/*.h",
+    ]) + [
+        "lib/xmltok_impl.c",
+        "lib/xmltok_ns.c",
+    ],
+    copts = [
+        "-DHAVE_MEMMOVE",
+        "-DXML_POOR_ENTROPY",
+    ],
+    includes = [
+        "lib",
+    ],
+    visibility = ["//visibility:public"],
+    deps = [],
+)
--
2.17.2
```

Also see attachment:
[0001.patch.txt](https://github.com/tensorflow/io/files/3138644/0001.patch.txt)

**Comment by yongtang** [comment]
@zou000 With the above the build should pass.

**Comment by terrytangyuan** [comment]
@yongtang Perhaps let's get that patch PR'ed into master first. Otherwise the commit history here might be messed up and we might encounter issues with the CLA. WDYT?

**Comment by yongtang** [comment]
@zou000 The PR #210 has been merged. You can rebase the master and make the following changes in libaprutil1.BUILD:

```patch
diff --git a/third_party/libaprutil1.BUILD b/third_party/libaprutil1.BUILD
index 6545137..ad9b6ac 100644
--- a/third_party/libaprutil1.BUILD
+++ b/third_party/libaprutil1.BUILD
@@ -4,7 +4,10 @@ licenses(["notice"])  # Apache license (for libapr1)
 
 cc_library(
     name = "libaprutil1",
-    srcs = [
+    srcs = glob([
+        "include/private/apu_*.h",
+        "include/private/apr_*.h",
+    ]) + [
         "buckets/apr_brigade.c",
         "buckets/apr_buckets.c",
         "buckets/apr_buckets_alloc.c",
@@ -112,10 +115,10 @@ cc_library(
         "-lrt",
         "-lcrypt",
         "-lpthread",
-        "-lexpat",
     ],
     strip_include_prefix = "include",
     deps = [
         "@libapr1",
+        "@libexpat",
     ],
 )
```

**Comment by zou000** [comment]
> @zou000 The PR #210 has been merged. You can rebase the master and make the following changes in libaprutil1.BUILD:
Thanks @yongtang and @terrytangyuan for looking into this and adding libexpat library. I made the change and I think it should pass CI this time.

Like @yongtang mentioned above it would be great to have a development docker container consistent with CI and release environment. #203 introduced 2 changes that accidentally made OSS work. Besides the libexpat library, the image also contains a /etc/bazelrc file that circumvented a couple of bazel-in-docker bugs. @yongtang's suggested patch also circumvented this bug, but in a different way. Personally, I think the bazelrc way is better as it makes development the same in docker and on host. Though I will leave it to you guys to decide.

I will update #200 with the info and close it.

**Comment by yongtang** [comment]
@zou000 @terrytangyuan Looks like there are still some issues on certain platforms. I think you could remove `-lexpat` and `-lcrypt`, and add boringssl as the dependency

In other word, apply the following patch:
```patch
diff --git a/third_party/libaprutil1.BUILD b/third_party/libaprutil1.BUILD
index 038a361..3b0d5d0 100644
--- a/third_party/libaprutil1.BUILD
+++ b/third_party/libaprutil1.BUILD
@@ -113,13 +113,12 @@ cc_library(
     ],
     linkopts = [
         "-lrt",
-        "-lcrypt",
         "-lpthread",
-        "-lexpat",
     ],
     strip_include_prefix = "include",
     deps = [
         "@libapr1",
         "@libexpat",
+        "@boringssl//:ssl",
     ],
 )
```

**Comment by zou000** [comment]
Thanks and done. It would be great if we have a whitelist of allowed
dynamic libraries available on systems and a submit checker to enforce it.
-- Li

**Comment by yongtang** [comment]
@zou000 We are almost there. I managed to get MacOS building correctly with additional changes:
https://github.com/tensorflow/io/commit/2efaf487770daca1af7bc685f8088fd0c325d9fa

The Travis CI passed with the above commit applied on top of your PR:

https://travis-ci.org/yongtang/io/builds/527888963

Can you apply the changes in the commit and update the PR?

**Comment by zou000** [comment]
Thanks Yong. I was working on the same direction but using conditional
macros. It seems working in CI.
-- Li

**Comment by terrytangyuan** [comment]
Thanks for all the efforts!

**Pull request merged**